cdc/metadata.cc

```diff
@@ -0,0 +1,154 @@
+ * limitations under the License.
+ *
+ * Modified by ScyllaDB
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "dht/murmur3_partitioner.hh"
+
+#include "cdc/cdc.hh"
+#include "cdc/metadata.hh"
+
+extern logging::logger cdc_log;
+
+static api::timestamp_type to_ts(db_clock::time_point tp) {
+    // FIXME: this assumes that timestamp_clock and db_clock have the same epochs
+    return std::chrono::duration_cast<api::timestamp_clock::duration>(tp.time_since_epoch()).count();
+}
+
+cdc::stream_id get_stream(
+        const std::vector<cdc::stream_id>& streams,
+        uint64_t ignore_msb_bits,
+        dht::token tok) {
+    // The ith stream is the stream for the ith shard.
+    auto shard_cnt = streams.size();
+    auto shard_id = dht::murmur3_partitioner(shard_cnt, ignore_msb_bits).shard_of(tok);
+
+    assert(shard_id < shard_cnt);
+
+    return streams[shard_id];
+}
+
+cdc::stream_id get_stream(
+        const std::vector<dht::token>& ranges,
+        const std::vector<std::vector<cdc::stream_id>>& streams,
+        const std::vector<uint64_t>& ignore_msb_bits,
+        dht::token tok) {
+    assert(!ranges.empty());
+    assert(streams.size() == ranges.size() && ignore_msb_bits.size() == ranges.size());
+
+    auto it = std::lower_bound(ranges.begin(), ranges.end(), tok);
+    if (it == ranges.end()) {
+        it = ranges.begin();
+    }
+
+    auto ix = it - ranges.begin();
+    return get_stream(streams[ix], ignore_msb_bits[ix], tok);
+}
+
+cdc::metadata::container_t::const_iterator cdc::metadata::get_current_gen(api::timestamp_type now) const {
+    auto it = _gens.upper_bound(now);
+    if (it == _gens.begin()) {
+        // All known generations have higher timestamps than `now`.
+        return _gens.end();
+    }
+
+    return std::prev(it);
+}
+
+cdc::stream_id cdc::metadata::get_stream(api::timestamp_type ts, dht::token tok) {
+    auto now = api::new_timestamp();
+    if (ts > now) {
+        throw exceptions::invalid_request_exception(format(
+                "cdc: attempted to get a stream \"from the future\" ({}; current time: {})."
+                " With CDC you cannot send writes with timestamps in the future, because we don't"
+                " know what streams will be used at that time.", print_ts(ts), print_ts(now)));
```

Comment:
`get_stream` is called from the coordinator of the write. The timestamp comes either from the coordinator or from the user. If it comes from the coordinator, there won't be any problem (assuming that its clock only moves forward).
If it comes from the user, then it's the user's responsibility to choose a timestamp in the past (but not earlier than the current generation's timestamp).

Reply:
Ok, I digged into this issue further and you're right.
The python driver, for instance, by default uses client-generated timestamps for writes. And cqlsh uses the python driver.

And anyway we should support a use case where the client chooses timestamps, at least if they are not too weird.

I'll fix it and give it a little leeway (a few seconds in the future max). It will be a little harder later to maintain consistency when we use transactions to make sure that generations are separated and no write goes to the wrong transaction, but it is doable.

Reply:
I added the ability to send writes into the near future